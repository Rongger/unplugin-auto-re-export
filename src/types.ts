import { __FILENAME__ } from "./parse";

export type Dir = string | string[];

export interface Options {
  dir: Dir;
  ignore?: string[];
  outputFile?: string;
}

export interface Exports {
  default?: string | typeof __FILENAME__;
  nameds: string[];
}
