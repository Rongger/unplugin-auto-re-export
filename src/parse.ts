import type { Exports } from "./types";
import * as t from "@babel/types";
import { parse } from "@babel/parser";
import traverse from "@babel/traverse";
import path from "path";
import fs from "fs";
import humps from "humps";
import type { TraverseOptions } from "babel__traverse";
import type { ParseResult } from "@babel/parser";

export const __FILENAME__ = Symbol(`__FILENAME__`);

export function parseExport(filePath: string): Exports {
  const ast = generateAST(filePath);

  return traverseAST(ast);
}

export function generateRaw(
  exports: Exports,
  dirPath: string,
  filePath: string
) {
  const def =
    exports.default === __FILENAME__
      ? humps.camelize(path.parse(filePath).name)
      : exports.default;

  const exportList = (exports.default ? [`default as ${def}`] : []).concat(
    exports.nameds
  );

  if (exportList.length === 0) return "";

  const relativePath = getRelativePath(dirPath, filePath);
  return `export { ${exportList.join(", ")} } from "${relativePath}";`;
}

export function generateAST(filePath: string) {
  const content = fs.readFileSync(filePath, "utf-8");
  const ast = parse(content, {
    sourceType: "module",
    plugins: ["typescript"],
  });
  return ast;
}

export function traverseAST(ast: ParseResult<t.File>): Exports {
  const namedExports: string[] = [];
  let defaultExport;

  const visitor: TraverseOptions = {
    ExportNamedDeclaration(path) {
      if (!path.node.declaration) return;

      const name = parseDeclaration(path.node.declaration);
      if (name) namedExports.push(name as string);
    },
    ExportDefaultDeclaration(path) {
      const name = parseDeclaration(path.node.declaration);
      if (name) defaultExport = name;
    },
    ExportSpecifier(path) {
      if (t.isIdentifier(path.node.exported))
        path.node.exported.name === "default"
          ? (defaultExport = path.node.local.name)
          : namedExports.push(path.node.exported.name);
    },
  };

  traverse(ast, visitor);

  return {
    nameds: namedExports,
    default: defaultExport,
  };
}

function getRelativePath(dirPath: string, filePath: string) {
  const { dir, name } = path.parse(filePath);

  return "./" + path.join(path.relative(dirPath, dir), name);
}

function parseDeclaration(declaration: t.Declaration | t.Expression) {
  let name;

  if (t.isIdentifier(declaration)) name = declaration.name;
  // 函数导出
  else if (t.isFunctionDeclaration(declaration)) {
    if (declaration.id) name = declaration.id.name;
    else name = __FILENAME__;
  }
  // 变量导出
  else if (t.isVariableDeclaration(declaration)) {
    declaration.declarations.forEach((d) => {
      // 解构导出
      if (t.isObjectPattern(d.id)) {
        d.id.properties.forEach((property) => {
          if (t.isObjectProperty(property) && t.isIdentifier(property.value)) {
            name = property.value.name;
          }
        });
      }
      // 导出单个特性
      else if (t.isIdentifier(d.id)) name = d.id.name;
    });
  }
  // 类导出
  else if (t.isClassDeclaration(declaration)) {
    name = declaration.id.name;
  }

  return name;
}
