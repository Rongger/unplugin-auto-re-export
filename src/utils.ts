import type { Dir, Options, OptionsMap, MapValue } from "./types";
import fg from "fast-glob";
import fs from "fs";
import path from "path";
import { exec } from "child_process";
import { parseExport, generateRaw, generateExportAllRaw } from "./parse";

const timerMap: Map<string, NodeJS.Timeout> = new Map();

export function genReExportFile(options: Options, path: string) {
  const dirs = resolveDir(options.dir);
  const result = dirs.find((i: string) => path.startsWith(i));
  if (!result) return;

  // 防止同一文件夹下的多个文件同时更改引发不必要的重复生成 index
  if (timerMap.has(result)) clearTimeout(timerMap.get(result));
  timerMap.set(
    result,
    setTimeout(() => {
      writeExportFromDir(result, options);
    }, 500)
  );
}

export function writeExportFromDir(dirPath: string, options: Options) {
  const [indexPath] = getOutputFilePaths(dirPath, options.outputFile);
  const optionsMap = genOptionsMap(options);
  const exportAll = getOptionsValue(optionsMap, dirPath, "exportAll");
  const baseNameMatch = getOptionsValue(optionsMap, dirPath, "baseNameMatch");

  const files = fg.sync(
    `${dirPath}/**/${baseNameMatch}.{js,jsx,ts,tsx,mjs,cjs}`,
    {
      ignore: [...options.ignore, indexPath],
      deep: getOptionsValue(optionsMap, dirPath, "deep"),
    }
  );

  const content = files.reduce<string>((raws, path) => {
    const raw = exportAll
      ? generateExportAllRaw(dirPath, path)
      : generateRaw(parseExport(path), dirPath, path);

    if (raw) raws += `${raw}\n`;
    return raws;
  }, "");

  if (content) {
    fs.writeFileSync(indexPath, content);
    tryToFormatCode(indexPath);
  }
}

export function resolveDir(dir: Dir) {
  if (typeof dir === "string") return [dir];
  return dir.map((i) => (typeof i === "string" ? i : i.path));
}

export function getOutputFilePaths(dir: Dir, outputFile: string) {
  return resolveDir(dir).map((dirPath) => path.join(dirPath, outputFile));
}

export const resolveDefaultOptions = ({
  dir = [],
  ignore = [],
  outputFile = "index.js",
  exportAll = false,
  deep = Infinity,
  baseNameMatch = "*",
  ...arg
}: Partial<Options>): Options => {
  return { dir, ignore, outputFile, exportAll, deep, baseNameMatch, ...arg };
};

export function genOptionsMap(options: Options) {
  const { dir, exportAll, deep, baseNameMatch } = options;
  const map: OptionsMap = new Map([
    ["DEFAULT", { exportAll, deep, baseNameMatch }],
  ]);
  if (Array.isArray(dir)) {
    dir.forEach((i) => {
      if (typeof i !== "string") {
        const {
          exportAll = options.exportAll,
          deep = options.deep,
          baseNameMatch = options.baseNameMatch,
        } = i;
        map.set(i.path, { exportAll, deep, baseNameMatch });
      }
    });
  }
  return map;
}

export function getOptionsValue<T extends keyof MapValue<OptionsMap>>(
  map: OptionsMap,
  dirPath: string,
  key: T
) {
  return (map.has(dirPath) ? map.get(dirPath) : map.get("DEFAULT"))![key];
}

function tryToFormatCode(codePath: string) {
  exec(`eslint --fix  ${codePath}`);
  exec(`prettier --write ${codePath}`);
}
