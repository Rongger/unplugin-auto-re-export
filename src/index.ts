import { createUnplugin } from "unplugin";
import chokidar from "chokidar";
import type { Options } from "./types";
import { genReExportFile } from "./utils";

export default createUnplugin<Options>((options) => {
  let watcher;
  if (options) {
    watcher = chokidar.watch(options.dir, {
      persistent: true,
    });
    watcher
      .on("add", (path) => genReExportFile(options, path))
      .on("change", (path) => genReExportFile(options, path))
      .on("unlink", (path) => genReExportFile(options, path));
  }

  return {
    name: "unplugin-auto-re-export",
    // buildEnd() {
    //   watcher.close().then(() => console.log("closed"));
    // },
  };
});
