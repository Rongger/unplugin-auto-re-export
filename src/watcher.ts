import chokidar from "chokidar";
import { resolveDir, genReExportFile, getOutputFilePaths } from "./utils";
import type { Options } from "./types";

export function initWatcher(options: Options) {
  let watcher;

  if (options.dir) {
    watcher = chokidar.watch(resolveDir(options.dir), {
      persistent: true,
      ignored: [
        ...options.ignore,
        ...getOutputFilePaths(options.dir, options.outputFile),
      ],
    });
    watcher
      .on("add", (path) => genReExportFile(options, path))
      .on("change", (path) => genReExportFile(options, path))
      .on("unlink", (path) => genReExportFile(options, path));
  }
  return watcher;
}
