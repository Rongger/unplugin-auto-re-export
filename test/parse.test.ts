import path from "path";
import { describe, expect, test } from "vitest";
import { parseExport, generateRaw } from "../src/parse";

describe("parse", () => {
  const dirPath = path.resolve(__dirname, "./fixtures");

  test("parseExport", async () => {
    expect(parseExport(path.join(dirPath, "/exports.ts"))).toMatchSnapshot();
  });

  test("parseTSExport", async () => {
    expect(parseExport(path.join(dirPath, "/ts-exports.ts"))).toMatchSnapshot();
  });

  test("generateRaw", () => {
    expect(
      generateRaw(
        {
          default: "l",
          nameds: [
            "utilB",
            "a",
            "b",
            "d",
            "ClassName",
            "e",
            "f",
            "g1",
            "h2",
            "bar",
          ],
        },
        dirPath,
        path.join(dirPath, "/exports.ts")
      )
    ).toMatchSnapshot();
  });
});
