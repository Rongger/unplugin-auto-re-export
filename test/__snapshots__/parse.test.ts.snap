// Vitest Snapshot v1

exports[`parse > generateExportAllRaw 1`] = `"export * from \\"./exports\\";"`;

exports[`parse > generateRaw 1`] = `"export { default as l, utilB, a, b, d, ClassName, e, f, g1, h2, bar } from \\"./exports\\";"`;

exports[`parse > parseExport 1`] = `
{
  "default": "l",
  "nameds": [
    "utilB",
    "a",
    "b",
    "d",
    "ClassName",
    "e",
    "f",
    "g1",
    "h2",
    "bar",
  ],
}
`;

exports[`parse > parseTSExport 1`] = `
{
  "default": undefined,
  "nameds": [
    "Type",
    "StringValidator",
    "ZipCodeValidator",
    "mainValidator",
    "T",
  ],
}
`;
